ntbs-service/Middleware/MsOfficeLinkPrefetchMiddleware.cs

```diff
@@ -0,0 +1,88 @@
+    // anti-CSRF correlation cookie set before the redirect is lost, and when the user is redirected back from Azure AD
+    // this cookie is missing, causing an app exception and a 500 response.
+
+    // Because the initial redirected request gets a successful response (302), MS Office then chooses to open the site
+    // properly in the browser. This leads to two tabs being opened, and one of which with the app working normally, and
+    // the other is an error 500 page.
+
+    // This issue has been present for over a decade. Discussion of this with regards to OIDC in ASP.NET is here:
+    // https://github.com/aspnet/Security/issues/1252
+    // And Microsoft's explanation for the behaviour is here:
+    // https://docs.microsoft.com/en-US/office/troubleshoot/office-suite-issues/click-hyperlink-to-sso-website
+    // Both parties suggest that other should fix the problem, so it seems unlikely that it will be fixed soon.
+    // This behaviour in MS Office can be turned off by editing a registry value, but we can't ask all our users to do
+    // this.
+
+    // To work around this, this middleware intercepts requests that come from MS Office, and sends back an empty
+    // 200 responses. When a user follows a hyperlink in MS Office, the initial MS Office request gets intercepted
+    // here, so no authorisation redirects are sent back (which would lead the second 500 tab). MS Office then asks the
+    // web browser to open the same URL, and it works how the user expected. The requests are intercepted based on the
+    // user-agent header.
+
+    // This class is taken from this StackOverflow answer for this problem
+    // https://stackoverflow.com/a/52478746
+    public class MsOfficeLinkPrefetchMiddleware
+    {
+        readonly RequestDelegate _next;
+
+        public MsOfficeLinkPrefetchMiddleware(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+        public Task Invoke(HttpContext context)
+        {
+            if (Is(context, HttpMethod.Get, HttpMethod.Head) && IsMsOffice(context))
+            {
+                // Mitigate by preempting auth challenges to MS Office apps' preflight requests and
+                // let the real browser start at the original URL and handle all redirects and cookies.
+
+                // Success response indicates to Office that the link is OK.
+                context.Response.StatusCode = (int)HttpStatusCode.OK;
+                context.Response.Headers["Cache-Control"] = "no-cache, no-store, must-revalidate";
+                context.Response.Headers["Pragma"] = "no-cache";
+                context.Response.Headers["Expires"] = "0";
+            }
+            else if (_next != null)
+            {
+                return _next.Invoke(context);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static bool Is(HttpContext context, params HttpMethod[] methods)
```

Comment:
I think `Is` or maybe `IsOneOf` would be an ok name if this were an extension method on the `context`, but as a helper method it's not very easy to understand. Maybe `RequestMethodMatches` would be a better name?